mysys/my_largepage.c

```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
-      fprintf(stderr,
-              "Warning: Failed to allocate %lu bytes from HugeTLB memory."
-              " errno %d\n", (ulong) size, errno);
+    mapflag= MAP_PRIVATE | MAP_ANONYMOUS;
+    if (my_use_large_pages)
+    {
+      large_page_size= my_next_large_page_size(*size, &page_i);
+      if (large_page_size)
+      {
+#ifdef HAVE_LINUX_LARGE_PAGES
+        mapflag|= MAP_HUGETLB | my_bit_log2(large_page_size) << MAP_HUGE_SHIFT;
+#else
+        mapflag|= MAP_ALIGNED_SUPER | MAP_ALIGNED(my_bit_log2(large_page_size));
+#endif
+      }
+    }
+    /* mmap adjusts the size to the hugetlb page size so no adjustment is needed */
+    ptr= mmap(NULL, *size, PROT_READ | PROT_WRITE, mapflag, -1, 0);
+    if (ptr == (void*) -1)
+    {
+      ptr= NULL;
+      if (my_flags & MY_WME)
+      {
+        if (large_page_size)
+        {
+          fprintf(stderr,
+                  "Warning: Failed to allocate %zu bytes from HugeTLB memory"
+                  "(page size %zu). errno %d\n", *size, large_page_size, errno);
+        }
+        else
+        {
+          fprintf(stderr,
+                  "Warning: Failed to allocate %zu bytes from memory."
+                  " errno %d\n", *size, errno);
+        }
+      }
+      /* try next smaller memory size */
+      if (large_page_size && errno == ENOMEM)
+        continue;
 
-    DBUG_RETURN(NULL);
+      /* other errors are more serious */
+      DBUG_RETURN(NULL);
+    }
+    else /* success */
+    {
+      if (large_page_size)
+      {
+         /* we do need to record the adjustment so that munmap gets called with
+            the right size. This is only the case for HUGETLB pages.
+            Align block size to large_page_size */
+         *size = MY_ALIGN(*size, (size_t) large_page_size);
+      }
+      DBUG_RETURN(ptr);
+    }
+    if (large_page_size == 0)
+    {
+      break; /* no more options to try */
+    }
   }
+  DBUG_RETURN(ptr);
+}
+
+#endif /* defined(HAVE_LINUX_LARGE_PAGES) || defined(HAVE_MMAP_ALIGNED) */
+
+#if defined(HAVE_GETPAGESIZES) && !defined(HAVE_LINUX_LARGE_PAGES)
+void my_get_large_page_sizes(size_t sizes[my_large_page_sizes_length])
+{
+  int nelem;
+
+  nelem = getpagesizes(NULL, 0);
 
-  ptr = (uchar*) shmat(shmid, NULL, 0);
-  if (ptr == (uchar *) -1)
+  assert(nelem <= my_large_page_sizes_length);
+  getpagesizes(sizes, my_large_page_sizes_length);
+  qsort(sizes, nelem, sizeof(size_t), size_t_cmp);
+  if (nelem < my_large_page_sizes_length)
   {
-    if (my_flags& MY_WME)
-      fprintf(stderr, "Warning: Failed to attach shared memory segment,"
-              " errno %d\n", errno);
-    shmctl(shmid, IPC_RMID, &buf);
+    sizes[nelem]= 0;
+  }
+}
+#endif
 
-    DBUG_RETURN(NULL);
+#if defined(HAVE_MMAP) && !defined(_WIN32)
+
+/* mmap and Linux-specific large pages deallocator */
+
+my_bool my_large_free_int(void *ptr, size_t size)
+{
+  DBUG_ENTER("my_large_free_int");
+
+# if defined(UNIV_SOLARIS)
+  if (munmap(static_cast<caddr_t>(ptr), size))
+# else
```

Comment:
IIUC it was needed for some ancient Solaris'es. Now they should happily accept `void*`.